### PR TITLE
Retrieve method for _initialized member of JApplication

### PR DESCRIPTION
**Pull request by andrea-celentano**
Adding a method to retrieve the private member _initialized from a JApplication object. Also, in the Initialize() method, put the _initialized=true line at the end, just before reutnr

### CODE REVIEW

**Comment by nathanwbrei** [comment]
I don't mind these changes, but I'm very curious as to why you needed them. 

1. Moving `_initialized=true` to the end is the right thing to do, but it suggests that you are experiencing exceptions thrown from inside `JApplication::Initialize()` and trying to recover from them. Up until now we've `exit()`ed the program under such circumstances, and I wouldn't be surprised if the `JApplication` is left in an inconsistent state. I just need to know if this is what motivated that. We will need to correctly recover from exceptions inside `JApplication::Initialize()` pretty soon regardless.

2. We should probably make `_initialized` be a `std::once_flag` and call `Initialize()` via `std::call_once` instead of reimplementing that functionality badly.

3. I'm really curious why you need `IsInitialized`, particularly since JApplication::Run will call `JApplication::Initialize` for you if you haven't. I had imagined you'd create a JApplication and launch a new thread which merely calls `JApplication::Run`. Do you have multiple threads which are testing for initialization and/or trying to call Initialize?

**Comment by andrea-celentano** [comment]
Hi Nathan,
thanks for your comment! The reason is the following - I attach the code I am using since it is the simplest way to see it.

This is the code that is called on the plugin. It is called for every group of events to be processed. Unfortunately, there isn’t any method that is called once at the beginning, hence I am using this “isFirst” flag.

The reason why I am waiting for the app to be initialized is that I call the “Run” method on a different thread, and it may take some time to initialize, if cal. constants are read from a DB for example. Hence, I put the blocking loop at the end. Maybe not necessary?

Thanks!

Andrea




static thread_local int isFirst = 1;
static thread_local JApplication * app = new JApplication();
static thread_local TridasEventSource* evt_src = new TridasEventSource("blocking_source", app);

if (isFirst) {
		std::cout << "TrigJANA isFirst being called" << std::endl;
		isFirst = 0;
		std::cout << "Adding the event source to the Japplication" << std::endl;
		app->Add(evt_src);
		std::cout << "DONE" << std::endl;

		std::cout << "Adding the factories Generators"<<std::endl;
		addRecoFactoriesGenerators(app);
		std::cout << " Done "<<std::endl;


		std::cout << "Adding the event processor the Japplication" << std::endl;
		app->Add(new GroupedEventProcessor());
		std::cout << "DONE" << std::endl;

		//Launch the thread with the JApplication and detach from it.
		std::cout << "TrigJANA starting the thread" << std::endl;
		boost::thread janaThread(JANAthreadFun, app);
		janaThread.detach();
		std::cout << "DONE and DETACHED" << std::endl;
		while (!app->IsInitialized()) {
			std::this_thread::sleep_for(std::chrono::milliseconds(10));
		}
	}


The thread fun reads:

void JANAthreadFun(JApplication *app) {
	std::cout << "TrigJANA JANAthreadFun start" << std::endl;
	app->Run();
	std::cout << "TrigJANA JANAthreadFun end" << std::endl;
}

**Comment by nathanwbrei** [comment]
Hi Andrea, I just merged this. Thank you for showing me how you are using it. That looks reasonable. There's a pretty strong chance that the code would behave correctly if you simply omitted the 
```		
while (!app->IsInitialized()) {
			std::this_thread::sleep_for(std::chrono::milliseconds(10));
		}
``` 
but keeping it in is fine, too. We might even be able to make it more concise: 

`JApplication::Run()` takes a parameter `wait_until_finished`, defaulted to true. If you set that to false, the method becomes non-blocking: it fires up the worker threads and immediately exits while the worker threads carry on. The upside is that your code is simpler and you have one fewer thread contending for a core. The downside is that you won't have the ticker announcing your progress and throughput, although you can bring the ticker back in your own code by simply calling `JApplication::PrintStatus` in your own code. I am not sure that this functionality actually works because we've never needed it, so it hasn't been fully tested, but I definitely want it demonstrably working at some point. Anyway, the (hypothetical) code would look like this:

```
if (isFirst) {
		std::cout << "TrigJANA isFirst being called" << std::endl;
		isFirst = 0;
		std::cout << "Adding the event source to the Japplication" << std::endl;
		app->Add(evt_src);
		std::cout << "DONE" << std::endl;

		std::cout << "Adding the factories Generators"<<std::endl;
		addRecoFactoriesGenerators(app);
		std::cout << " Done "<<std::endl;


		std::cout << "Adding the event processor the Japplication" << std::endl;
		app->Add(new GroupedEventProcessor());
		std::cout << "DONE" << std::endl;

		std::cout << "Starting JANA" << std::endl;
		app->Run(false);
	}
```

**Comment by andrea-celentano** [comment]
Hi Nathan,
thanks! I’ll try the app->Run(false) method, it can simplify the code by having one thread less - and this is definitively desirable!

Bests,
Andrea

**Comment by nathanwbrei** [comment]
There is one other issue I see in your code listing, by the way -- You've declared isFirst, app, and evt_src to be `static thread_local`. What that is doing is creating a separate variable visible only to that thread. What you want is to declare them simply `static`, since only one instance of JApplication and evt_src are meant to be shared among both threads. Six months ago having two instances of JApplication would have caused things to fail in horrible ways, so I'm quite happy that it didn't. However, there are certain weird corners where I still expect JANA to break (for instance, with the JCalibrationManager I just re-added) so I still recommend changing it.

**Comment by nathanwbrei** [comment]
Hi Andrea,
Great, let me know how it goes! And if it misbehaves, just send me the details and don't spend too much time poking at it. Meanwhile I'll be cleaning up JCalibrationManager.

**Comment by andrea-celentano** [comment]
Hi Nathan,
the reason why I set these to be thread_local is that the void TrigJANA(PluginArgs const& args) function may be called by different threads in the same process (the TriDAS TCPU process), and I thought that, to speed up things, I preferred to have a separate JApplication for each thread.

You mention in your previous email: 

>  What you want is to declare them simply static, since only one instance of JApplication and evt_src are meant to be shared among both threads.
> 
What are the two threads you are referring to?

Thanks
Andrea

**Comment by nathanwbrei** [comment]
When I said 'two threads', I hadn't realized you had multiple threads each running `TrigJANA`, so I meant the master thread and the `JANAThreadFun` thread (whose local `app` parameter shadows the static thread-local `app` variable). Your clarification makes more sense, but keep in mind that each `JApplication` instance manages its own thread pool with (by default) as many threads as cores, so you should get better performance if all threads reference the same `JApplication` instance.

**Comment by andrea-celentano** [comment]
Hi Nathan,
thanks. May I ask you to explain me (your C++ is stronger than mine!):
> so I meant the master thread and the JANAThreadFun thread (whose local app parameter shadows the static thread-local app variable).
> 

The JApplication is created in the TrigJANA function (one per thread running it), and then a pointer to it is passed to JANAthreadFun via:

boost::thread janaThread(JANAthreadFun, app);

so that the janaThread function acts as a “server” to keep the same JApplication running (this was before I knew JApplication can be non-blocking).

Also, concerning the idea of having the JApplication not as thread_local, I have the following question. When the first thread calls TrigJANA the first time, it enters the “isFirst” if, where the JApplication is initialized (japp->Run()). What if another, different thread enters in TrigJANA at the almost the same time, and tries to “feed” events to the JApplication when this is not yet initialized?

Thanks!
Andrea
